design-proposals/vhostuser/vhostuser-interface.md

```diff
@@ -0,0 +1,166 @@
+
+- With the above process a kubevirt VM with a vHostUser interface can be acheived. All the steps highlighted will be the changes made as part of the design.
+
+## API Examples
+
+Since, KubeVirt will always explicitly define the pod interface name for multus-cni. It will be computed from the VMI spec interface name, to allow multiple connections to the same multus provided network.
+
+The vHostUser Interface will be defined in the VM spec as  shown below:
+
+```yaml
+          interfaces:
+          - name: vhost-user-vn-blue
+            vhostuser: {}
+          useVirtioTransitional: true
+      networks:
+      - name: vhost-user-vn-blue
+        multus:
+          networkName: vn-blue
+```
+No additional definition of vloumes or annotations are required. The DownwardAPI will provide all the necessary annotations required for vhostUser Interface.
+
+## Scalability
+- There should not be any scalability issues, as the feature is to create an interface when useful.
+- VM migration works the same way, as the only change is having a new interace, which only requires a dataplane of the host in userpsace.
+
+## Update/Rollback Compatibility
+- Should have no impact.
+
+## Functional Testing Approach
+Functional test can:
+
+- Create VM with vhostUser interface
+- Create VM with multiple interface types. (Bridge + vhostuser type)
+
+# Implementation Phases
+- The design will be implemented in two phases:
+    - Add DPDK support on kubevirt/kubevirtci
+        - Enable OVS with DPDK
+        - Enable an option in gocli for DPDK support
+        - Add UserSpace CNI in the setup instead of or along with calico.
+    - Add vhostUser Interface type in kubevirt/kubevirt
+        - Add vHostUser Interface type
+        - Create Pod template to support the Interface
+        - Create appropriate virsh xml elements
+        - Add E2E test to send traffic between the 2 VMs created.
+
+## Annex
+
+
+A sample DownwardAPi mght look this way:
+
+```yaml
+k8s.v1.cni.cncf.io/network-status="[{\n    \"name\": \"k8s-kubemanager-kubernetes-CNI/default-podnetwork\",\n    \"interface\": \"eth0\",\n    \"ips\": [\n        \"10.244.2.4\"\n    ],\n    \"mac\": \"02:d7:b3:27:b3:b2\",\n    \"default\": true,\n    \"dns\": {}\n},{\n    \"name\": \"kubevirttest/vn-blue\",\n    \"interface\": \"net1\",\n    \"ips\": [\n        \"19.1.1.2\"\n    ],\n    \"mac\": \"02:c6:da:d3:ab:07\",\n    \"dns\": {},\n    \"device-info\": {\n        \"type\": \"vhost-user\",\n        \"vhost-user\": {\n            \"mode\": \"server\",\n            \"path\": \"2ea1931c-2935-net1\"\n        }\n    }\n}]"
+k8s.v1.cni.cncf.io/networks="[{\"interface\":\"net1\",\"name\":\"vn-blue\",\"namespace\":\"kubevirttest\",\"cni-args\":{\"interface-type\":\"virtio\"}}]"
```

Comment:
What does `"interface-type":"virtio"` mean? Is that related to vhost-user-net or is net1 exposed as a virtio-net device somehow?

Reply:
If the user is trying to create a vhostUser interface, we are passing the interface type as virtio in the multus cni arguments in the multusNetworkAnnotation from kubevirt code. This is the reason we see that in the pod annotations.

Reply:
Why "virtio" and not "vhost-user-net"?

Reply:
It might have come up there as we use that name in out CNI. We can change it to host-user-net and have it as a standard.

Reply:
I have updated the cni arguments in the annotations. This value might not be used by every CNI, however any CNI wanted to use the CNI arguments can use this as a standard if needed.

Reply:
Thanks. "virtio" or "virtio-net" can mean something different from "vhost-user-net" (e.g. a virtio-net PCI or vdpa-net device). Calling it "vhost-user-net" is clearer.